libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2046,22 +2046,31 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 		}
 		string errorMsg = "Member \"" + memberName + "\" not found or not visible "
 				"after argument-dependent lookup in " + exprType->toString() + ".";
-		if (memberName == "value")
-		{
-			errorMsg.pop_back();
-			errorMsg +=	" - did you forget the \"payable\" modifier?";
-		}
-		else if (exprType->category() == Type::Category::Function)
+
+		if (exprType->category() == Type::Category::Function)
 		{
 			if (auto const& funType = dynamic_pointer_cast<FunctionType const>(exprType))
 			{
 				auto const& t = funType->returnParameterTypes();
+
+				bool isStruct = false;
+				bool isContract = false;
+
 				if (t.size() == 1)
-					if (
-						t.front()->category() == Type::Category::Contract ||
-						t.front()->category() == Type::Category::Struct
-					)
-						errorMsg += " Did you intend to call the function?";
+				{
+					isStruct = t.front()->category() == Type::Category::Struct;
+					isContract = t.front()->category() == Type::Category::Contract;
+				}
+
+				if (memberName == "value")
+				{
+					if (isContract)
+						errorMsg = "Constructor for " + t.front()->toString() + " must be payable for member \"value\" to be available.";
```

Comment:
I think this error message is only correct in a special case. Consider the following test:
```
contract C {
  function f() public view returns (C) { return this; }
  function g() { this.f.value(); }
}
```
I think in the current version, this would yield the message about the constructor, which would be wrong.

Reply:
I think you have to query `funType->kind == FunctionType::Kind::Creation`